server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java

```diff
@@ -454,8 +454,14 @@ private static boolean isUnassignedDueToTimelyRestart(ShardRouting routing, Node
         return now - restartingAllocationDelayExpiration <= 0;
     }
 
-    private static boolean isUnassignedDueToNewInitialization(ShardRouting routing) {
-        return routing.primary() && routing.active() == false && getInactivePrimaryHealth(routing) == ClusterHealthStatus.YELLOW;
+    private static boolean isUnassignedDueToNewInitialization(ShardRouting routing, ClusterState state) {
+        if (routing.active()) {
+            return false;
+        }
+        // If the primary is inactive for unexceptional events in the cluster lifecycle, both the primary and the
+        // replica are considered new initializations.
+        ShardRouting primary = routing.primary() ? routing : state.routingTable().shardRoutingTable(routing.shardId()).primaryShard();
+        return primary.active() == false && getInactivePrimaryHealth(primary) == ClusterHealthStatus.YELLOW;
```

Comment:
Let's examine this case by case, I do not think all of them should be covered here. 

> a primary of new index just started and corresponding replica is initialising for the first time (should not degrade health)

This is true, but then we need to distinguish between creating and just initialising from other reasons. Because currently, initialising replicas for other reasons are meant to be `yellow` the same way that initialising primaries turn `red`. Is my thinking clear?

Reply:
Correct, the issue that I see is following:
```
index created, both primary and replica are unassigned: status is green (was yellow prior this change)
primary is initializing, replica remain unassigned: status is green (was yellow prior this change)
primary is started, replica is unassigned: status is yellow
primary is started, replica is initializing: status is yellow
primary and replica are started: status is green
```

With this change the health would still change to yellow during index creation even though for a shorter period of time.

I agree that it would really help to have a clear indication that the shard is initializing for the first time especially as replicas are always initializing the same way.

Reply:
Ok, so in order to separate these cases, we need to know that the primary just `STARTED`, but without keeping history, I do not think we can do that right?

If indeed history is necessary, then we would have to accept for now at least, that this is something we cannot distinguish, and momentarily we might get a yellow status. Do you agree?

Reply:
Thanks for the conversation Ievgen and Mary. ++ on what has been said and advised so far. 

Just out of curiosity, do we require a history in any of the remaining cases? 

Say we have:
```
my-index 0 p STARTED
my-index 0 r INITIALIZING
```
Does it matter if the primay _just_ started or has been started for a while?

Reply:
@andreidan , we might want to differentiate between a replica that for lost and it's recovering and a replica that is starting up for the first time either because the user created an index or increased the number of replicas. The second one could be thought of that there is no impact since we just now requested the change while the first one could be indicating a problem if the replica was lost because of an issue.

Reply:
++ that makes sense thanks Mary 

Note that we do have support for looking at `unassignedInfo` when evaluating health however, we only do so in verbose mode as it's a rather expensive operation, especially in the many-shards scenario.